File: harness/lib/saas-core/python/domain/tenant_aware_entity.py
from __future__ import annotations

from abc import ABC
from datetime import datetime, timezone

from .tenant_id import TenantId


class TenantAwareEntity(ABC):
    def __init__(self, id: str, tenant_id: str, created_at: datetime | None = None) -> None:
        if not id:
            raise ValueError(f"{self.__class__.__name__}: id is required")
        if not tenant_id:
            raise ValueError(f"{self.__class__.__name__}: tenant_id is required")
        self.id = id
        self.tenant_id = tenant_id
        self.created_at = created_at or datetime.now(tz=timezone.utc)

    def get_tenant_id(self) -> TenantId:
        return TenantId.from_string(self.tenant_id)

    def belongs_to(self, tenant_id: str) -> bool:
        return self.tenant_id == tenant_id

File: harness/lib/saas-core/node/domain/tenant-aware-entity.ts
import { TenantId } from './tenant-id';

export abstract class TenantAwareEntity {
  protected constructor(
    public readonly id: string,
    public readonly tenantId: string,
    public readonly createdAt: Date = new Date(),
  ) {
    if (!id) throw new Error(`${this.constructor.name}: id is required`);
    if (!tenantId) throw new Error(`${this.constructor.name}: tenantId is required`);
  }

  getTenantId(): TenantId {
    return TenantId.fromString(this.tenantId);
  }

  belongsTo(tenantId: string): boolean {
    return this.tenantId === tenantId;
  }

  toJSON(): Record<string, unknown> {
    return {
      id: this.id,
      tenantId: this.tenantId,
      createdAt: this.createdAt.toISOString(),
    };
  }
}
